### DISCUSSION TITLE
How to navigate to a screen stack inside a tab navigator?

### DISCUSSION THREAD

**Discussion by xmflsct** [react-navigation/react-navigation#9217] [Q&A]
Say I have below setup:
```
Tab-1
    Stack
        Screen -1(root, default)
Tab-2
    Stack
        Screen-1 (root, default)
        Screen-2 (logically next level of Screen-1, like settings sub page)
        Screen-3 (logically next level of Screen-2, like settings sub sub page)
```

Using `navigation.navigate()` I am able to get from `Tab-1 Screen-1` to any screens inside `Tab-2`. But since logically the screens inside `Tab-2` are nested, think like phone's settings. If I navigate to `Tab-2 Screen-3` directly the first time, when I try to go back (go back button is cusomized therefore it is there even it cannot go back), I cannot logically go back from `Screen-3` to `Screen-2` then to `Screen-1`.

How can I achieve such navigation pattern? Thanks!

**Comment by raajnadar** [comment]
Just answer some question I will try to help you.

Can we navigate directly to Tab-2 -> Screen-3?
Is Screen-1, Screen-2 & Screen-3 are siblings or nested?

If possible add a snack demo it will be easy for me to understand the scenario.

**Reply by xmflsct** [reply]
Yes `Tab-2 -> Screen-3` works without problem.
All 3 screens inside `Tab-2` are siblings within a `Stack` like mentioned above. But logically they are not siblings but hierarchical/nested.

You can try the live app here https://github.com/tooot-app/app.

Or like in iOS's system settings, `General -> About -> Name`.
What I want to achieve is that, I can navigate to page `Name` directly, and then when using `goBack()`, it shall go back to `About`. I can hardcode it with a `navigate('About')` but then the animation is incorrect.

**Reply by raajnadar** [reply]
For that you need to use the `navigation.reset()` API to design a navigation tree yourself. 

https://reactnavigation.org/docs/navigation-prop/#reset

https://reactnavigation.org/docs/navigation-state

Please try this and let me know.

**Reply by cabiad** [reply]
This is a much later reply. Probably not much use to @xmflsct, but there seems to be [at least one open issue on this topic](#8036) where people are having issues generating the correct `reset()` call in v6 when trying to jump between navigators.

So, for anyone else like me, coming along later with a similar problem, here's what I did to complete @raajnadar's suggestion above:

1. Manually navigate to the screen in the destination navigator and log or otherwise view the output of `navigation.getState()` while it's open. (If logging, use JSON.stringify() to see the content of the nested objects too, they are important.)
2. Remove everything from that state object except:
* the hierarchy of screen objects and navigators, 
* the `index`, `routes`, and `name`, `state`, and `params` values from each level in the hierarchy
* Be careful to replace any `params` values (like object IDs) with the correct current value
3. Use this as the value of the `state` param passed to `navigation.reset()`.

This got me a working `reset()` call!

From this experience, with nested navigators, it was easy to miss that the navigation state passed to `reset()` seems to need information about _the state of each of the navigators in the hierarchy and not just the list of screens themselves_. The [reset docs](https://reactnavigation.org/docs/navigation-prop/#reset) were a bit confusing on this for me.

I wonder if the developers would be interested in adding a guide to generating state arguments to `reset()` in more complex situations, or if something in the API could be tweaked so this use case (jump between navigators) can be more easily supported.

Lastly, note that the docs suggest that everything but `index` and `routes` might change in minor versions of react navigation, so if any changes have happened, others might need additional or different parts of the state if you're using a different minor version than me.
